### PR TITLE
Allow cell sizes > 1.0 for linked cells, argument validation, refactoring

### PR DESCRIPTION
**Pull request by Garfield96**
# Description

1. Allow bigger cell sizes (>1.0) for linked cells container. Cell size option was added to md-flexible as well.
2. Cutoff and cell size are now checked in setters of autopas.h to avoid invalid values.
3. CMake: Copy script files to build dir to avoid manual copying.
4. Refactoring of testing helpers.

# How Has This Been Tested?

1. Tests were added to CellBlock3dTest to test new cell sizes.
3. AutoPasTest.cpp tests whether invalid arguments causes an exception.

### CODE REVIEW

File: examples/md-flexible/measurePerf.sh [hunk 1]
```diff
@@ -81,73 +81,87 @@ do
...
-
-                # workaround because there is no traversal for Verlet clusters with newton 3 yet.
-                if [[ ${container} =~ 'VerletCluster' && ${newton3Opt} =~ 'on' ]];
-                then
-                    continue
-                fi
-
-                # iterate over molecules with the correct repetition
-                for i in `seq 0 $(( ${#Mols[@]} - 1 ))` ;
+                for cellSize in 1.0  1.25  1.5  1.75  2.0  2.5  3.0;
```
Comment (FG-TUM) [inline examples/md-flexible/measurePerf.sh]: Testing multiple cell sizes does not make sense for all containers. Restrict this only to LinkedCells for now.
Also, I think 7 values are a bit excessive :D Choose 3 which you think are interesting.

File: tests/testAutopas/testingHelpers/GaussianGenerator.h [hunk 2]
```diff
@@ -11,21 +11,36 @@
 
 class GaussianGenerator {
  public:
+  /**
+   * fills a autopas object with gaussian distributed particles
```
Comment (FG-TUM) [inline tests/testAutopas/testingHelpers/GaussianGenerator.h]:
```suggestion
   * Fills an autopas object with Gaussian distributed particles.
```

File: tests/testAutopas/testingHelpers/GridGenerator.h [hunk 3]
```diff
@@ -19,44 +19,46 @@ class GridGenerator {
...
-                                const Particle &defaultParicle = autopas::Particle(),
-                                std::array<double, 3> spacing = std::array<double, 3>{1, 1, 1},
-                                std::array<double, 3> offset = std::array<double, 3>{.5, .5, .5});
+  static void fillWithParticles(std::vector<ParticleCell>& cells, const std::array<size_t, 3>& particlesPerDim,
+                                const Particle& defaultParticle = autopas::Particle(),
+                                const std::array<double, 3>& spacing = {1.0, 1.0, 1.0},
+                                const std::array<double, 3>& offset = {.5, .5, .5});
 
   /**
    * fills a autopas object with a cuboid mesh of particles
```
Comment (FG-TUM) [inline tests/testAutopas/testingHelpers/GridGenerator.h]:
```suggestion
   * Fills an autopas object with a cuboid mesh of particles.
```

File: tests/testAutopas/testingHelpers/RandomGenerator.h [hunk 4]
```diff
@@ -11,60 +11,106 @@
 
 class RandomGenerator {
  private:
+  /**
+   * returns a random number between fMin and fMax
```
Comment (FG-TUM) [inline tests/testAutopas/testingHelpers/RandomGenerator.h]:
```suggestion
   * Returns a random number between fMin and fMax.
```

File: tests/testAutopas/testingHelpers/RandomGenerator.h [hunk 5]
```diff
@@ -11,60 +11,106 @@
...
+   * @param fMin min. number (inclusive)
+   * @param fMax max. number (inclusive)
+   * @return double
+   */
   static double fRand(double fMin, double fMax);
 
  public:
+  /**
+   * returns a random position in the space between boxMin and boxMax
```
Comment (FG-TUM) [inline tests/testAutopas/testingHelpers/RandomGenerator.h]:
```suggestion
   * Returns a random position in the space between boxMin and boxMax.
```

File: tests/testAutopas/testingHelpers/RandomGenerator.h [hunk 6]
```diff
@@ -11,60 +11,106 @@
...
+   * @param boxMin min. position
+   * @param boxMax max. position
+   * @return std::array<double, 3>
+   */
   static std::array<double, 3> randomPosition(const std::array<double, 3>& boxMin, const std::array<double, 3>& boxMax);
 
+  /**
+   * fills the given container with random distributed particles
```
Comment (FG-TUM) [inline tests/testAutopas/testingHelpers/RandomGenerator.h]:
```suggestion
   * Fills the given container with randomly distributed particles.
```

File: tests/testAutopas/testingHelpers/RandomGenerator.h [hunk 7]
```diff
@@ -11,60 +11,106 @@
...
 
-    for (unsigned long i = 0; i < numParticles; ++i) {
-      Particle particle = defaultParticle;
-      particle.setR(randomPosition(container.getBoxMin(), container.getBoxMax()));
-      particle.setID(i);
-      container.addParticle(particle);
-    }
-  }
+  /**
+   * fills the given container with random distributed particles between boxMin and boxMax
```
Comment (FG-TUM) [inline tests/testAutopas/testingHelpers/RandomGenerator.h]:
```suggestion
   * Fills the given container with randomly distributed particles between boxMin and boxMax.
```

File: tests/testAutopas/testingHelpers/RandomGenerator.h [hunk 8]
```diff
@@ -11,60 +11,106 @@
...
+   * @param boxMax max. position
+   * @param numParticles number of particles
+   */
+  template <class Container, class Particle>
+  static void fillWithParticles(Container& container, const Particle& defaultParticle,
+                                const std::array<double, 3>& boxMin, const std::array<double, 3>& boxMax,
+                                unsigned long numParticles = 100ul);
 
+  /**
+   * fills halo of given container with n = numParticles particles
```
Comment (FG-TUM) [inline tests/testAutopas/testingHelpers/RandomGenerator.h]:
```suggestion
   * Fills the halo of the given container with n = numParticles particles.
```

File: examples/md-flexible/measurePerf.sh [hunk 9]
```diff
@@ -81,73 +81,87 @@ do
...
                 do
-                    thisReps=${Reps[$i]}
-                    # Direct sum is slow for huge number of particles -> limit number of iterations
-                    if [[ ${container} = 'DirectSum' && ${Mols[$i]} -ge 2048 ]];
+                    separate "Cell size ${cellSize}"
+
+                    configPrinted=false
+
+                    filename="runtimes_${container}_${dataLayout}_N3${newton3Opt}_cs${cellSize}"
```
Comment (FG-TUM) [inline examples/md-flexible/measurePerf.sh]: here too

File: tests/testAutopas/tests/containers/CellBlock3DTest.h [hunk 10]
```diff
@@ -14,19 +14,23 @@ class CellBlock3DTest : public AutoPasTestBase {
  public:
   CellBlock3DTest()
       : _cells_1x1x1(_vec1, std::array<double, 3>({0.0, 0.0, 0.0}), std::array<double, 3>({10.0, 10.0, 10.0}), 10.0),
+        _cells_1x1x1_cs(_vec1_cs, std::array<double, 3>({0.0, 0.0, 0.0}), std::array<double, 3>({10.0, 10.0, 10.0}),
```
Comment (FG-TUM) [inline tests/testAutopas/tests/containers/CellBlock3DTest.h]:
```suggestion
        _cells_1x1x1_cs2(_vec1_cs, std::array<double, 3>({0.0, 0.0, 0.0}), std::array<double, 3>({10.0, 10.0, 10.0}),
```
Also adapt the other names

File: src/autopas/containers/linkedCells/LinkedCells.h [hunk 11]
```diff
@@ -37,11 +37,13 @@ class LinkedCells : public ParticleContainer<Particle, ParticleCell, SoAArraysTy
    * @param boxMin
    * @param boxMax
    * @param cutoff
+   * @param cellSize
```
Comment (FG-TUM) [inline src/autopas/containers/linkedCells/LinkedCells.h]: State here again that this is a factor and not directly the absolute size

File: src/autopas/containers/CellBlock3D.h [hunk 12]
```diff
@@ -34,11 +34,12 @@ class CellBlock3D : public CellBorderAndFlagManager {
    * @param vec vector of ParticleCells that this class manages
    * @param bMin lower corner of the cellblock
    * @param bMax higher corner of the cellblock
-   * @param interactionLength minimal size of each cell
+   * @param interactionLength max. radius of interaction between particles
+   * @param cellSize cell size ralative to interactionLength
```
Comment (SteffenSeckler) [inline src/autopas/containers/CellBlock3D.h]:
```suggestion
   * @param cellSize cell size relative to interactionLength
```

Comment (Garfield96) [reply src/autopas/containers/CellBlock3D.h]: I'll change the name from cellSize to cellSizeFactor to show that this is a relative value.
> Additionally, this factor should always be bigger than 1.0, as a decreased cell size is normally not allowed.
>
Regarding this PR, this is correct, but support for smaller cell sizes is coming soon (see branch linked-cell-smaller-cell-sizes).

**Comment by SteffenSeckler** [comment]
> > Additionally, this factor should always be bigger than 1.0, as a decreased cell size is normally not allowed.
> 
> Regarding this PR, this is correct, but support for smaller cell sizes is coming soon (see branch linked-cell-smaller-cell-sizes).

hmm ok, could you document that this only affects the linked cell versions?

**Comment by SteffenSeckler** [comment]
in case you do not need the branch anymore, please delete it :)

**Pull request merged**